### PR TITLE
introduce strategy factory

### PR DESCRIPTION
**Pull request by shubham-cmyk**
This PR is part of https://github.com/kyverno/kuttl/issues/23

### CODE REVIEW

File: pkg/test/utils/subset.go [hunk 1]
```diff
@@ -49,17 +54,27 @@ func IsSubset(expected, actual interface{}) error {
 
 	switch reflect.TypeOf(expected).Kind() {
 	case reflect.Slice:
-		if reflect.ValueOf(expected).Len() != reflect.ValueOf(actual).Len() {
-			return &SubsetError{
-				message: fmt.Sprintf("slice length mismatch: %d != %d", reflect.ValueOf(expected).Len(), reflect.ValueOf(actual).Len()),
-			}
+		var strategy ArrayComparisonStrategy
+		if strategyFactory != nil {
```
Comment (eddycharly) [inline pkg/test/utils/subset.go:58]: this should not be nil

File: pkg/test/utils/subset.go [hunk 2]
```diff
@@ -49,17 +54,27 @@ func IsSubset(expected, actual interface{}) error {
...
+
+		expectedData := make([]interface{}, expectedVal.Len())
+		actualData := make([]interface{}, actualVal.Len())
+
+		for i := 0; i < expectedVal.Len(); i++ {
+			expectedData[i] = expectedVal.Index(i).Interface()
 		}
+		for i := 0; i < actualVal.Len(); i++ {
+			actualData[i] = actualVal.Index(i).Interface()
+		}
```
Comment (eddycharly) [inline pkg/test/utils/subset.go:75]: let's not duplicate the same code twice

File: pkg/test/utils/subset.go (hunk 1 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go:58]: this could be to pass some test case I have passed that nil

File: pkg/test/utils/subset.go (hunk 2 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go:75]: are we sure the length of acutal and expected to be same

File: pkg/test/utils/subset.go [hunk 3]
```diff
@@ -49,17 +54,27 @@ func IsSubset(expected, actual interface{}) error {
...
 
-		for i := 0; i < reflect.ValueOf(expected).Len(); i++ {
-			if err := IsSubset(reflect.ValueOf(expected).Index(i).Interface(), reflect.ValueOf(actual).Index(i).Interface()); err != nil {
-				return err
-			}
+		expectedVal := reflect.ValueOf(expected)
+		actualVal := reflect.ValueOf(actual)
+
+		expectedData := make([]interface{}, expectedVal.Len())
+		actualData := make([]interface{}, actualVal.Len())
```
Comment (eddycharly) [inline pkg/test/utils/subset.go:68]: an optim would be to let the strategy handle this if needed (example, if the len are different, no need to copy)

File: pkg/test/utils/subset.go (hunk 2 above)
Comment (eddycharly) [reply pkg/test/utils/subset.go:75]: why do we care ?

File: pkg/test/utils/subset.go (hunk 1 above)
Comment (eddycharly) [reply pkg/test/utils/subset.go:58]: what ?

File: pkg/test/utils/subset.go (hunk 2 above)
Comment (eddycharly) [reply pkg/test/utils/subset.go:75]:
```
		toSlice := func(v interface{}) []interface{} {
			value := reflect.ValueOf(v)
			slice := make([]interface{}, value.Len())
			for i := 0; i < value.Len(); i++ {
				slice[i] = value.Index(i).Interface()
			}
			return slice
		}
		return strategy(toSlice(expected), toSlice(actual))
```

File: pkg/test/utils/subset.go (hunk 1 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go:58]: I mean currently the error is also using IsSubset but we have not implemented factory over there for a time can there be a nil check ?

File: pkg/test/utils/subset.go (hunk 2 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go:75]: looks good

File: pkg/test/utils/subset.go (hunk 1 above)
Comment (eddycharly) [reply pkg/test/utils/subset.go:58]: create a `defaultFactory` eventually

File: pkg/test/utils/subset.go (hunk 2 above)
Comment (eddycharly) [reply pkg/test/utils/subset.go:75]: still, there's https://github.com/kyverno/kuttl/pull/39/files#r1336389395

Comment (shubham-cmyk) [reply pkg/test/utils/subset.go]: fixed

File: pkg/test/utils/subset.go (hunk 1 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go:58]: so would that use the matchExact ?

File: pkg/test/utils/subset.go (hunk 2 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go]: can you check again ?

File: pkg/test/utils/subset.go (hunk 3 above)
Comment (eddycharly) [reply pkg/test/utils/subset.go]: no resolved

File: pkg/test/utils/subset.go [hunk 4]
```diff
@@ -5,6 +5,11 @@ import (
 	"reflect"
 )
 
+type ArrayComparisonStrategyFactory func(path string) ArrayComparisonStrategy
+type ArrayComparisonStrategy func(expectedData, actualData []interface{}) error
```
Comment (eddycharly) [inline pkg/test/utils/subset.go:9]:
```suggestion
type ArrayComparisonStrategyFactory = func(path string) ArrayComparisonStrategy
type ArrayComparisonStrategy = func(expectedData, actualData []interface{}) error
```

File: pkg/test/utils/subset.go [hunk 5]
```diff
@@ -5,6 +5,11 @@ import (
...
+
+var StrategyFactory ArrayComparisonStrategyFactory
```
Comment (eddycharly) [inline pkg/test/utils/subset.go:11]: what's that ?

Comment (shubham-cmyk) [reply pkg/test/utils/subset.go:11]: IsSubet called from : https://github.com/kyverno/kuttl/pull/39/files#diff-33f37fadcc7b670b7b50c39f28ca1b37c37ffc59da7870dc7bfcea5e7942529dR104

How are they gonna know about the stratergyFactory

Comment (eddycharly) [reply pkg/test/utils/subset.go:11]:
```suggestion
fun defaultFactory (path string) ArrayComparisonStrategy {
  return StrategyExact(path)
}
```

File: pkg/test/utils/subset.go (hunk 1 above)
Comment (eddycharly) [reply pkg/test/utils/subset.go:58]: see https://github.com/kyverno/kuttl/pull/39/files#r1336407308

File: pkg/test/utils/subset.go [hunk 6]
```diff
@@ -49,17 +54,15 @@ func IsSubset(expected, actual interface{}) error {
 
 	switch reflect.TypeOf(expected).Kind() {
 	case reflect.Slice:
-		if reflect.ValueOf(expected).Len() != reflect.ValueOf(actual).Len() {
-			return &SubsetError{
-				message: fmt.Sprintf("slice length mismatch: %d != %d", reflect.ValueOf(expected).Len(), reflect.ValueOf(actual).Len()),
-			}
+		var strategy ArrayComparisonStrategy
```
Comment (eddycharly) [inline pkg/test/utils/subset.go:57]:
```suggestion
if strategyFactory == nil {
  strategyFactory = defaultFactory
}
strategy := strategyFactory(currentPath)
```

File: pkg/test/utils/subset.go (hunk 3 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go]: now fixed

File: pkg/test/utils/subset.go [hunk 7]
```diff
@@ -90,3 +94,54 @@ func IsSubset(expected, actual interface{}) error {
...
+
+func StrategyExact(path string) ArrayComparisonStrategy {
+	return func(expected, actual interface{}) error {
+
+		if reflect.ValueOf(expected).Len() != reflect.ValueOf(actual).Len() {
+			return &SubsetError{message: fmt.Sprintf("slice length mismatch at path %s: %d != %d", path, reflect.ValueOf(expected).Len(), reflect.ValueOf(actual).Len())}
+		}
+
+		expectedData := toSlice(expected)
+		actualData := toSlice(actual)
```
Comment (eddycharly) [inline pkg/test/utils/subset.go:128]: you don't need to copy anymore :-)

File: pkg/test/utils/subset.go [hunk 8]
```diff
@@ -90,3 +94,54 @@ func IsSubset(expected, actual interface{}) error {
...
+
+		for i, v := range expectedData {
+			newPath := path + fmt.Sprintf("[%d]", i)
+			if err := IsSubset(v, actualData[i], newPath, StrategyFactory); err != nil {
```
Comment (eddycharly) [inline pkg/test/utils/subset.go:132]: I didn't notice the global `StrategyFactory` here, we can't do that, it's not thread safe

File: pkg/test/utils/subset.go (hunk 3 above)
Comment (eddycharly) [reply pkg/test/utils/subset.go]: 👍

File: pkg/test/utils/subset.go (hunk 5 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go:11]: this could lead to initialization cycle

Comment (shubham-cmyk) [reply pkg/test/utils/subset.go:11]: coud I use : 

```
	case reflect.Slice:
		if strategyFactory == nil {
			strategyFactory = StrategyExact
		}
		strategy := strategyFactory(currentPath)
```

File: pkg/test/utils/subset.go (hunk 8 above)
Comment (shubham-cmyk) [reply pkg/test/utils/subset.go]: fixed now

**Pull request merged**